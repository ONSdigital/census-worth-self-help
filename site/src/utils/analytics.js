export default class Analytics {
  static trackEvent(category, action, name = "", value = "") {
    if (window._paq) {
      window._paq.push(["trackEvent", category, action, name, value])
    } else {
      console.debug(`Track event not set : ${category} - ${action} - ${name} - ${value}`)
    }
  }

  static trackSiteSearch(keyword, category, searchCount) {
    if (window._paq) {
      window._paq.push(["trackSiteSearch", keyword, category, searchCount])
    } else {
      console.debug(`Track site search : ${keyword} - ${category} - ${searchCount}`)
    }
  }

  static setPageType(pageType) {
    if (typeof window !== `undefined` && window._paq) {
      window._paq.push(["setCustomDimension", 1, pageType])
    }
  }
}