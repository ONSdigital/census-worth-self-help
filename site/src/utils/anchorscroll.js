export default (history, timeout = 1000) => {
  let observer
  let timeoutId

  const reset = () => {
    if (timeoutId) {
      clearTimeout(timeoutId)
      timeoutId = null
    }

    if (observer) {
      observer.disconnect()
    }
  }

  const convertTextToLink = text => {
    return text.toLowerCase().replace(/\s/g, "-")
  }

  const createScrollToElement = hash => {
    return () => {
      let headers = []
      let tags = ["h1", "h2", "h3"]

      tags.forEach(tag => {
        headers = headers.concat(Array.from(document.getElementsByTagName(tag)))
      })

      let element = headers.find(
        header => convertTextToLink(header.textContent) === hash
      )

      if (element) {
        element.scrollIntoView()
        reset()
        return true
      }

      return false
    }
  }

  function scroll(hash) {
    if (typeof hash !== "string") {
      return
    }

    const anchorText = hash.slice(1).toLowerCase()

    if (!anchorText) {
      return
    }

    const scrollToElement = createScrollToElement(anchorText)

    setTimeout(() => {
      if (scrollToElement()) {
        return
      }

      if (window.MutationObserver) {
        observer = new MutationObserver(scrollToElement)

        observer.observe(document, {
          attributes: true,
          childList: true,
          subtree: true
        })
      }

      timeoutId = setTimeout(reset, timeout)
    })
  }

  // listen for changes
  history.listen(({ location, action }) => {
    if (timeoutId) {
      reset()
    }

    if (action !== "POP") {
      return
    }

    if (location.hash) {
      scroll(location.hash)
    }
  })

  // also trigger on load
  if (history.location.hash) {
    scroll(history.location.hash)
  }
}
