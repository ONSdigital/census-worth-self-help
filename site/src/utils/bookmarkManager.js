export default class BookmarkManager {
  constructor() {
    this.bookmarks = JSON.parse(localStorage.getItem("bookmarks"))
    if (this.bookmarks === null) {
      this.bookmarks = []
    }
  }

  getTopBookmarks() {
    return this.bookmarks
  }

  isPageBookmarked(page) {
    return this.bookmarks.includes(page)
  }

  bookmarkPage(page) {
    if (!this.isPageBookmarked(page)) {
      this.bookmarks.push(page)
    }
    localStorage.setItem("bookmarks", JSON.stringify(this.bookmarks))
  }

  unBookmarkPage(page) {
    this.bookmarks = this.bookmarks.filter(function(bookmark) {
      return bookmark !== page
    })

    localStorage.setItem("bookmarks", JSON.stringify(this.bookmarks))
  }
}
