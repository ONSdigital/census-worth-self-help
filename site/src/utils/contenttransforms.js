var sanitizeHtml = require("sanitize-html")

export const transformSources = htmlString => {
  let videoPath = process.env.GATSBY_ASSETS_PATH
    ? process.env.GATSBY_ASSETS_PATH
    : ""
  if (!htmlString) {
    return ""
  }
  return htmlString.replace(/{{TARGET_ASSETS_SRC}}/g, videoPath)
}

export const htmlSanitize = htmlString => {
  return sanitizeHtml(htmlString, {
    allowedTags: sanitizeHtml.defaults.allowedTags.filter(function(tag) {
      return tag !== "iframe"
    }).concat([
      "video",
      "audio",
      "img",
      "source",
      "h1",
      "h2"
    ]),
    allowedAttributes: {
      a: ["href", "name", "target"],
      img: ["src", "title"],
      source: ["src", "type"],
      video: ["controls"],
      audio: ["controls"]
    }
  })
}
