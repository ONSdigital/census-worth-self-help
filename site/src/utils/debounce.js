import _ from "lodash"

// debounce with default delay, or return sync function if delay exactly zero (for unit tests)
export default (fn, delay) => {
  return (delay === 0)
    ? fn
    : _.debounce((...args) => fn(...args), delay || 1000)
}