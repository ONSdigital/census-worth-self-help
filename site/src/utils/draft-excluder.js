const READY_FOR_PRODUCTION = "Ready for Live Site"

// Articles in draft should not be deployed to production. An article is draft if the draftreason for the article
// is anything other than "Ready for production" - see ONS-359.
//
// We also use this Draft Excluder concept to not show articles that are for the call centre only.
//
// Initially this was done using the draft tag. This technique might be removed in the future.
exports.excludeDraftArticle = node => {
  return Boolean(
    process.env.EXCLUDE_DRAFTS && (
      (
        node.frontmatter.tags &&
        node.frontmatter.tags.includes("draft")
      ) || (
        node.frontmatter.draftreason &&
        node.frontmatter.draftreason !== READY_FOR_PRODUCTION
      ) || node.frontmatter.cconly
    )
  )
}
