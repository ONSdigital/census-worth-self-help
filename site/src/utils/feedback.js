export default class Feedback {
  static bookmarkClickEvent(article) {
    Feedback.feedbackEvent("click-event", "bookmark", article)
  }
  static articleWasBookmarked(article) {
    Feedback.feedbackEvent("article-was-bookmarked", "bookmarked", article, 1)
  }
  static articleWasUnBookmarked(article) {
    Feedback.feedbackEvent("article-was-bookmarked", "unbookmarked", article, 1)
  }
  static articleIsUseful(article, reason) {
    Feedback.feedbackEvent(
        "article-feedback-review",
        "POSITIVE REVIEW: " + article,
        reason
    )
    Feedback.feedbackEvent("article-feedback-rating", "rating", article, 1)
  }
  static articleIsNotUseful(article, reason) {
    Feedback.feedbackEvent(
      "article-feedback-review",
      "REVIEW: " + article,
      reason
    )
    Feedback.feedbackEvent("article-feedback-rating", "rating", article, -1)
  }
  static feedbackEvent(category, action, name = "", value = "") {
    if (window._paq) {
      window._paq.push(["trackEvent", category, action, name, value])
    }
  }
}
