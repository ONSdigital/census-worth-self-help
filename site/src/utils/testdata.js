export const data = {
  edges: [
    {
      node: {
        frontmatter: {
          title: "test Article 1",
          description: "testDescription",
          priority: 1,
          date: new Date()
        },
        fields: {
          pagename: "test-article-1",
          collection: "articles"
        }
      }
    },
    {
      node: {
        frontmatter: {
          title: "test Article 2",
          description: "testDescription",
          priority: 0,
          date: new Date(),
          tags: ["popular", "webchat"]
        },
        fields: {
          pagename: "test-article-2",
          collection: "articles"
        }
      }
    },
    {
      node: {
        frontmatter: {
          title: "test Article 3",
          description: "testDescription",
          priority: 0,
          date: new Date(),
          tags: ["popular"]
        },
        fields: {
          pagename: "test-article-3",
          collection: "articles"
        }
      }
    },
    {
      node: {
        frontmatter: {
          title: "test Article 4",
          description: "testDescription",
          priority: 0,
          date: new Date(),
          tags: ["draft"]
        },
        fields: {
          pagename: "test-article-4",
          collection: "articles"
        }
      }
    },
    {
      node: {
        frontmatter: {
          title: "test Article 5",
          description: "testDescription",
          priority: 0,
          date: new Date(),
          tags: ["popular"]
        },
        fields: {
          pagename: "test-article-5",
          collection: "articles"
        }
      }
    },
    {
      node: {
        frontmatter: {
          title: "test Article 6",
          description: "testDescription",
          priority: 0,
          date: new Date(),
          tags: ["popular"]
        },
        fields: {
          pagename: "test-article-6",
          collection: "articles"
        }
      }
    },
    {
      node: {
        frontmatter: {
          title: "test Directory 7",
          description: "testDescription",
          priority: 0,
          date: new Date()
        },
        fields: {
          pagename: "test-directory-1",
          collection: "directories"
        }
      }
    }
  ]
}

export const articleNode = data.edges[0].node
export const webchatArticleNode = data.edges[1].node
export const directoryNode = data.edges[2].node
export const draftNode = data.edges[3].node
export const popularList = {
  edges: [data.edges[1], data.edges[2], data.edges[4], data.edges[5]]
}
export const articleList = { edges: data.edges.slice(0, 6) }
