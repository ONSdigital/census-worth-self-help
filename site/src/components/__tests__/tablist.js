import React from "react"
import renderer from "react-test-renderer"

import TabList from "../tablist"
import {data} from "../../utils/testdata"

describe("TabList", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(<TabList elements={data.edges} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})