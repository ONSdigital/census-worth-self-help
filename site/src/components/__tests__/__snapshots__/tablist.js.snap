// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabList renders correctly 1`] = `
<div
  css={
    Object {
      "map": undefined,
      "name": "1p7jpnp",
      "next": undefined,
      "styles": "
        
    padding: 15px 0px;
  ;
      ",
    }
  }
>
  <div>
    <div
      css={
        Object {
          "map": undefined,
          "name": "lcxyxv",
          "next": undefined,
          "styles": "
        
    padding: 15px;
    min-height: 30px;
  ;
        box-shadow: 0px 2px 4px 0px rgba(0, 0, 0, 0.5);
        border-left: 6px solid rgb(0,61,89);
        background-color: rgb(255, 255, 255);
      ",
        }
      }
      onClick={[Function]}
    >
      <div
        className="Card-heading-Style"
        css={
          Object {
            "map": undefined,
            "name": "qiy98z",
            "next": undefined,
            "styles": "
          color: inherit;
          flex-grow: 1;
          font-weight: semi-bold;
        ",
          }
        }
        data-testid="articletab-article-card"
      >
        test Article 1
      </div>
      <div
        className="Card-sub-head-Style-gray"
        css={
          Object {
            "map": undefined,
            "name": "1hejrb1",
            "next": undefined,
            "styles": "
          padding-top: 6px;
          display: flex;
          flex-direction: vertical;
          line-height: 1.08;
          max-height: 42.12px; // line-height * font-size * 3
        ",
          }
        }
      >
        <div
          css={
            Object {
              "map": undefined,
              "name": "lz6rtp",
              "next": undefined,
              "styles": "
            width: 70%;
            overflow: hidden;
            text-overflow: ellipsis;
            flex-grow: 1;
          ",
            }
          }
        >
          testDescription
        </div>
        <div
          className="Card-meta-Style"
          css={
            Object {
              "map": undefined,
              "name": "4j3tb8",
              "next": undefined,
              "styles": "
            display: flex;
            flex-direction: column;
            justify-content: flex-end;
          ",
            }
          }
        >
          a few seconds ago
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      css={
        Object {
          "map": undefined,
          "name": "lcxyxv",
          "next": undefined,
          "styles": "
        
    padding: 15px;
    min-height: 30px;
  ;
        box-shadow: 0px 2px 4px 0px rgba(0, 0, 0, 0.5);
        border-left: 6px solid rgb(0,61,89);
        background-color: rgb(255, 255, 255);
      ",
        }
      }
      onClick={[Function]}
    >
      <div
        className="Card-heading-Style"
        css={
          Object {
            "map": undefined,
            "name": "qiy98z",
            "next": undefined,
            "styles": "
          color: inherit;
          flex-grow: 1;
          font-weight: semi-bold;
        ",
          }
        }
        data-testid="articletab-article-card"
      >
        test Article 2
      </div>
      <div
        className="Card-sub-head-Style-gray"
        css={
          Object {
            "map": undefined,
            "name": "1hejrb1",
            "next": undefined,
            "styles": "
          padding-top: 6px;
          display: flex;
          flex-direction: vertical;
          line-height: 1.08;
          max-height: 42.12px; // line-height * font-size * 3
        ",
          }
        }
      >
        <div
          css={
            Object {
              "map": undefined,
              "name": "lz6rtp",
              "next": undefined,
              "styles": "
            width: 70%;
            overflow: hidden;
            text-overflow: ellipsis;
            flex-grow: 1;
          ",
            }
          }
        >
          testDescription
        </div>
        <div
          className="Card-meta-Style"
          css={
            Object {
              "map": undefined,
              "name": "4j3tb8",
              "next": undefined,
              "styles": "
            display: flex;
            flex-direction: column;
            justify-content: flex-end;
          ",
            }
          }
        >
          a few seconds ago
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      css={
        Object {
          "map": undefined,
          "name": "lcxyxv",
          "next": undefined,
          "styles": "
        
    padding: 15px;
    min-height: 30px;
  ;
        box-shadow: 0px 2px 4px 0px rgba(0, 0, 0, 0.5);
        border-left: 6px solid rgb(0,61,89);
        background-color: rgb(255, 255, 255);
      ",
        }
      }
      onClick={[Function]}
    >
      <div
        className="Card-heading-Style"
        css={
          Object {
            "map": undefined,
            "name": "qiy98z",
            "next": undefined,
            "styles": "
          color: inherit;
          flex-grow: 1;
          font-weight: semi-bold;
        ",
          }
        }
        data-testid="articletab-article-card"
      >
        test Article 3
      </div>
      <div
        className="Card-sub-head-Style-gray"
        css={
          Object {
            "map": undefined,
            "name": "1hejrb1",
            "next": undefined,
            "styles": "
          padding-top: 6px;
          display: flex;
          flex-direction: vertical;
          line-height: 1.08;
          max-height: 42.12px; // line-height * font-size * 3
        ",
          }
        }
      >
        <div
          css={
            Object {
              "map": undefined,
              "name": "lz6rtp",
              "next": undefined,
              "styles": "
            width: 70%;
            overflow: hidden;
            text-overflow: ellipsis;
            flex-grow: 1;
          ",
            }
          }
        >
          testDescription
        </div>
        <div
          className="Card-meta-Style"
          css={
            Object {
              "map": undefined,
              "name": "4j3tb8",
              "next": undefined,
              "styles": "
            display: flex;
            flex-direction: column;
            justify-content: flex-end;
          ",
            }
          }
        >
          a few seconds ago
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      css={
        Object {
          "map": undefined,
          "name": "lcxyxv",
          "next": undefined,
          "styles": "
        
    padding: 15px;
    min-height: 30px;
  ;
        box-shadow: 0px 2px 4px 0px rgba(0, 0, 0, 0.5);
        border-left: 6px solid rgb(0,61,89);
        background-color: rgb(255, 255, 255);
      ",
        }
      }
      onClick={[Function]}
    >
      <div
        className="Card-heading-Style"
        css={
          Object {
            "map": undefined,
            "name": "qiy98z",
            "next": undefined,
            "styles": "
          color: inherit;
          flex-grow: 1;
          font-weight: semi-bold;
        ",
          }
        }
        data-testid="articletab-article-card"
      >
        test Article 4
      </div>
      <div
        className="Card-sub-head-Style-gray"
        css={
          Object {
            "map": undefined,
            "name": "1hejrb1",
            "next": undefined,
            "styles": "
          padding-top: 6px;
          display: flex;
          flex-direction: vertical;
          line-height: 1.08;
          max-height: 42.12px; // line-height * font-size * 3
        ",
          }
        }
      >
        <div
          css={
            Object {
              "map": undefined,
              "name": "lz6rtp",
              "next": undefined,
              "styles": "
            width: 70%;
            overflow: hidden;
            text-overflow: ellipsis;
            flex-grow: 1;
          ",
            }
          }
        >
          testDescription
        </div>
        <div
          className="Card-meta-Style"
          css={
            Object {
              "map": undefined,
              "name": "4j3tb8",
              "next": undefined,
              "styles": "
            display: flex;
            flex-direction: column;
            justify-content: flex-end;
          ",
            }
          }
        >
          a few seconds ago
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      css={
        Object {
          "map": undefined,
          "name": "1rakub1",
          "next": undefined,
          "styles": "
        
    padding: 15px;
    min-height: 30px;
  ;
        box-shadow: 0 2px 4px 0 rgba(212, 212, 212, 0.5);
        border: solid 1px #ececed;
        
    background: linear-gradient(to left, rgb(25, 80, 105), rgb(71, 114, 136));
  ;
        border-left: 6px solid rgb(0,61,89);
      ",
        }
      }
      onClick={[Function]}
    >
      <div
        className="Card-heading-Style"
        css={
          Object {
            "map": undefined,
            "name": "1bl45q",
            "next": undefined,
            "styles": "
          color: white;
          flex-grow: 1;
          font-weight: semi-bold;
        ",
          }
        }
        data-testid="articletab-article-card"
      >
        test Directory 5
      </div>
      <div
        className="Card-sub-head-Style-gray"
        css={
          Object {
            "map": undefined,
            "name": "1hejrb1",
            "next": undefined,
            "styles": "
          padding-top: 6px;
          display: flex;
          flex-direction: vertical;
          line-height: 1.08;
          max-height: 42.12px; // line-height * font-size * 3
        ",
          }
        }
      >
        <div
          css={
            Object {
              "map": undefined,
              "name": "yv8zhf",
              "next": undefined,
              "styles": "
            color: white;
            width: 70%;
            overflow: hidden;
            text-overflow: ellipsis;
            flex-grow: 1;
          ",
            }
          }
        >
          testDescription
        </div>
      </div>
    </div>
  </div>
</div>
`;
